src/LtiAssignmentsGradesService.php

```diff
@@ -103,4 +118,40 @@ public function getGrades(LtiLineitem $lineitem)
 
         return $scores['body'];
     }
+
+    public function getLineItems()
+    {
+        if (!in_array(LtiConstants::AGS_SCOPE_LINEITEM, $this->service_data['scope'])) {
+            throw new LtiException('Missing required scope', 1);
+        }
+        $line_items = [];
+
+        $next_page = $this->service_data['lineitems'];
+
+        while ($next_page) {
+            $page = $this->service_connector->makeServiceRequest(
+                $this->service_data['scope'],
+                LtiServiceConnector::METHOD_GET,
+                $next_page,
+                null,
+                null,
+                'application/vnd.ims.lti-gs.v1.contextgroupcontainer+json'
+            );
+
+            $line_items = array_merge($line_items, $page['body']);
+            $next_page = false;
+            $link = $page['headers']['Link'];
+
+            if (preg_match(LtiServiceConnector::NEXT_PAGE_REGEX, $link, $matches)) {
+                $next_page = $matches[1];
+            }
+        }
```

Comment:
This logic is duplicated in multiple services. We should refactor it to a single place, maybe in the LtiServiceConnector.

Reply:
The logic is still wrong in the other places. I was considering that to be out of scope for this PR right now since they're already broken. I can handle refactoring in a subsequent PR